hathor/p2p/sync_v2/streamers.py

```diff
@@ -180,44 +184,67 @@ def send_next(self) -> None:
         # XXX: don't send the genesis or the current block
         if self.current_block is None or self.current_block.is_genesis:
             self.stop()
-            self.node_sync.send_blocks_end(StreamEnd.NO_MORE_BLOCKS)
+            self.sync_agent.send_blocks_end(StreamEnd.NO_MORE_BLOCKS)
             return
 
         self.schedule_if_needed()
 
 
 class TransactionsStreamingServer(_StreamingServerBase):
     """Streams all transactions confirmed by the given block, from right to left (decreasing timestamp).
+
+    If the start_from parameter is not empty, the BFS (Breadth-First Search) for the first block will commence
+    using the provided hashes. This mechanism enables streaming requests to continue from a specific point
+    should there be interruptions or issues.
     """
 
     def __init__(self,
-                 node_sync: 'NodeBlockSync',
+                 sync_agent: 'NodeBlockSync',
                  start_from: list[BaseTransaction],
                  first_block_hash: bytes,
                  last_block_hash: bytes,
                  *,
                  limit: int = DEFAULT_STREAMING_LIMIT) -> None:
         # XXX: is limit needed for tx streaming? Or let's always send all txs for
         # a block? Very unlikely we'll reach this limit
-        super().__init__(node_sync, limit=limit)
-
-        assert len(start_from) > 0
-        assert start_from[0].storage is not None
-        self.storage = start_from[0].storage
-        self.first_block_hash = first_block_hash
-        self.last_block_hash = last_block_hash
-        self.last_block_height = 0
-
-        self.bfs = BFSOrderWalk(self.storage, is_dag_verifications=True, is_dag_funds=True, is_left_to_right=False)
-        self.iter = self.bfs.run(start_from, skip_root=False)
+        super().__init__(sync_agent, limit=limit)
+
+        self.first_block: Block = cast(Block, self.tx_storage.get_transaction(first_block_hash))
+        self.last_block: Block = cast(Block, self.tx_storage.get_transaction(last_block_hash))
+        self.start_from = start_from
+
+        assert isinstance(self.first_block, Block)
+        assert isinstance(self.last_block, Block)
```

Comment:
If you assert before, then you don't need the cast above

Reply:
I need to improve this validation. I'm not sure whether I should assert it here and actually validate it on the caller or just validate it here. The point is that these hashes are sent by the other peer so we need to validate them somewhere.

Reply:
I see. The cast will only trick mypy, but the assert is the actual runtime validation. If it helps you could use `if not self.first_block.is_block: # close connection or something`.

Reply:
I want `self.first_block` and `self.last_block` to have type `Block`. But then the `self.tx_storage.get_transaction()` returns a `BaseTransaction` which causes mypy to fail. So I used a cast followed by an assert to make sure we have the right type (since cast does not actually check it).


I aim for `self.first_block` and `self.last_block` to specifically be typed as `Block`. However, the method `self.tx_storage.get_transaction()` returns a BaseTransaction, which causes mypy to fail. So, I initially used a cast to hint to mypy that the returned object should be treated as a `Block`. But since cast alone doesn't guarantee the object's type at runtime, I also added asserts to make sure that the objects are indeed instances of `Block`.

Reply:
I guess this verification should be made at the agent level. This assert just checks if the parameters were properly placed.

Reply:
Is this being done in this PR?

Reply:
It was partially implemented. It's complete now. https://github.com/HathorNetwork/hathor-core/pull/850/commits/35f8222bf3917636e72762f9375b3db9bc06052f

Reply:
I missed one validation. Here it is: https://github.com/HathorNetwork/hathor-core/pull/850/commits/00253b4b8482d698f028b2870348195763c368b9

Reply:
Ok!